src/jdk.compiler/share/classes/com/sun/tools/javac/launcher/MemoryContext.java

```diff
@@ -0,0 +1,194 @@
+        try {
+            List<File> searchPath = descriptor.fileObject().isFirstLineIgnored() ? List.of() : List.of(descriptor.sourceRootPath().toFile());
+            standardFileManager.setLocation(StandardLocation.SOURCE_PATH, searchPath);
+        } catch (IOException e) {
+            throw new Error("unexpected exception from file manager", e);
+        }
+        this.memoryFileManager = new MemoryFileManager(inMemoryClasses, standardFileManager);
+    }
+
+    String getSourceFileAsString() {
+        return descriptor.fileObject().getFile().toAbsolutePath().toString();
+    }
+
+    Set<String> getNamesOfCompiledClasses() {
+        return Set.copyOf(inMemoryClasses.keySet());
+    }
+
+    /**
+     * Compiles a source file, placing the class files in a map in memory.
+     * Any messages generated during compilation will be written to the stream
+     * provided when this object was created.
+     *
+     * @return the name of the first class found in the source file
+     * @throws Fault if any compilation errors occur, or if no class was found
+     */
+    String compileProgram() throws Fault {
+        var units = new ArrayList<JavaFileObject>();
+        units.add(descriptor.fileObject());
+        if (descriptor.isModular()) {
+            var root = descriptor.sourceRootPath();
+            units.add(standardFileManager.getJavaFileObject(root.resolve("module-info.java")));
+        }
+        var opts = options.forProgramCompilation();
+        var task = compiler.getTask(out, memoryFileManager, null, opts, null, units);
+        var fileUri = descriptor.fileObject().toUri();
+        var mainClassNameReference = new AtomicReference<String>();
+        task.addTaskListener(new TaskListener() {
+            @Override
+            public void started(TaskEvent event) {
+                if (mainClassNameReference.get() != null) return;
+                if (event.getKind() != TaskEvent.Kind.ANALYZE) return;
+                TypeElement element = event.getTypeElement();
+                if (element.getNestingKind() != NestingKind.TOP_LEVEL) return;
+                JavaFileObject source = event.getSourceFile();
+                if (source == null) return;
+                if (!source.toUri().equals(fileUri)) return;
+                var mainClassName = element.isUnnamed()
+                        ? element.getSimpleName().toString()
+                        : element.getQualifiedName().toString();
+                mainClassNameReference.compareAndSet(null, mainClassName);
+            }
+        });
+        var ok = task.call();
+        if (!ok) {
+            throw new Fault(Errors.CompilationFailed);
+        }
+        if (mainClassNameReference.get() == null) {
+            throw new Fault(Errors.NoClass);
+        }
+        return mainClassNameReference.get();
+    }
+
+    byte[] compileJavaFileByName(String name) throws Fault {
+        var firstDollarSign = name.indexOf('$'); // [package . ] name [ $ enclosed ]
```

Comment:
Nit - in principle, `$` is a valid identifier character:
```
jshell> Character.isJavaIdentifierStart('$')
$1 ==> true
````

So, in principle, the user's code may have classes with `$` in their names. OTOH, this seems like a very rare case given the context.